external-crates/move/tools/move-package/src/resolution/dependency_graph.rs

```diff
@@ -140,84 +126,706 @@ struct PackageWithResolverTOML<'a>(&'a Package);
+                .download_and_update_if_remote(
+                    *pkg_name,
+                    &internal_dep.kind,
+                    &mut self.progress_output,
+                )
+                .with_context(|| format!("Fetching '{}'", *pkg_name))?;
+            let pkg_path = root_path.join(local_path(&internal_dep.kind));
+
+            let Ok(lock_contents) = std::fs::read_to_string(pkg_path.join(SourcePackageLayout::Lock.path())) else {
+                return Ok(None);
+            };
+            hashed_lock_files.push(digest_str(lock_contents.as_bytes()));
+        }
+
+        Ok(if hashed_lock_files.is_empty() {
+            None
+        } else {
+            Some(hashed_lock_files)
+        })
+    }
+
+    /// Computes a digest of all dependencies in a manifest file but may return None if information
+    /// about some dependencies is not available.
+    fn dependency_digest(
+        &mut self,
+        root_path: PathBuf,
+        manifest: &PM::SourceManifest,
+    ) -> Result<Option<String>> {
+        let mut dep_hashes = self
+            .dependency_hashes(root_path.clone(), &manifest.dependencies)?
+            .or(Some(vec![]))
+            .unwrap();
+
+        let dev_dep_hashes = self
+            .dependency_hashes(root_path.clone(), &manifest.dev_dependencies)?
+            .or(Some(vec![]))
+            .unwrap();
+
+        if dep_hashes.is_empty() {
+            Ok(None)
+        } else {
+            dep_hashes.extend(dev_dep_hashes);
+            Ok(Some(hashed_files_digest(dep_hashes)))
+        }
+    }
+}
+
+impl DependencyGraph {
+    /// Finds packages in a sub-graph that should be pruned as a result of applying an override from
+    /// the outer graph. A package should be pruned if it's dominated by an overridden package.
+    fn find_pruned_pkgs(
+        &self,
+        pruned_pkgs: &mut BTreeMap<PM::PackageName, BTreeSet<PM::PackageName>>,
+        reachable_pkgs: &mut BTreeSet<PM::PackageName>,
+        root_pkg_name: PM::PackageName,
+        from_pkg_name: PM::PackageName,
+        mode: DependencyMode,
+        overrides: &BTreeMap<PM::PackageName, Package>,
+        dev_overrides: &BTreeMap<PM::PackageName, Package>,
+        overridden_path: bool,
+    ) -> Result<()> {
+        if overridden_path {
+            // we are on a path originating at the overridden package
+            if !reachable_pkgs.contains(&from_pkg_name) {
+                // not (yet) reached via regular (non-overridden) path - include in the list of
+                // packages to be pruned but need to include a list of neighbors to correctly remove
+                // edges going to the pruned package
+                let neighbors = BTreeSet::from_iter(
+                    self.package_graph
+                        .neighbors_directed(from_pkg_name, Direction::Incoming),
+                );
+                pruned_pkgs.insert(from_pkg_name, neighbors);
+            }
+        } else {
```

Comment:
The idea here was to keep a set of nodes that were visited on the non-overridden path and use it to control inclusion of nodes to be pruned:
- if a node is on the overridden path and not reachable via regular path, make it a pruning candidate
- if a node is discovered to be on the regular path, remove pruning this node from pruning candidate set if it has been made a candidate before

There was a bug in this logic before which I discovered when writing tests, but otherwise I think it should work...